compiler/p/codegen/ControlFlowEvaluator.cpp

```diff
@@ -1693,864 +2024,786 @@ TR::Register *OMR::Power::TreeEvaluator::ifacmpneEvaluator(TR::Node *node, TR::C
    {
-   TR::Register *trgReg    = cg->allocateRegister();
-   TR::Node     *firstChild   = node->getFirstChild();
-   TR::Node     *secondChild  = node->getSecondChild();
-   TR::Register *src1Reg = cg->evaluate(firstChild);
-   TR::Register *temp1Reg     = NULL;
-   TR::Register *temp2Reg     = cg->allocateRegister();
-   bool         killTemp1    = false;
-
-   if (secondChild->getOpCode().isLoadConst() &&
-       secondChild->getRegister() == NULL)
+   switch (cond)
       {
-      int64_t value = secondChild->get64bitIntegralValue();
-      if (value == 0)
+      case CompareCondition::lt:
          {
-         temp1Reg = src1Reg;
+         return srcReg;
          }
-      else
+
+      case CompareCondition::ge:
          {
-         temp1Reg = addConstantToInteger(node, src1Reg , -value, cg);
-         killTemp1 = true;
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::nor, node, tempReg, srcReg, srcReg);
+         return tempReg;
          }
+
+      case CompareCondition::gt:
+         {
+         generateTrg1Src1Instruction(cg, TR::InstOpCode::neg, node, tempReg, srcReg);
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::andc, node, tempReg, tempReg, srcReg);
+         return tempReg;
+         }
+
+      case CompareCondition::le:
+         {
+         generateTrg1Src1Instruction(cg, TR::InstOpCode::neg, node, tempReg, srcReg);
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::orc, node, tempReg, srcReg, tempReg);
+         return tempReg;
+         }
+
+      default:
+         TR_ASSERT_FATAL_WITH_NODE(node, false, "Invalid compare condition %d for intOrderZeroToSignBit", static_cast<int>(cond));
       }
-   else
-      {
-      temp1Reg = cg->allocateRegister();
-      TR::Register *src2Reg   = cg->evaluate(secondChild);
-      generateTrg1Src2Instruction(cg, TR::InstOpCode::subf, node, temp1Reg, src2Reg ,src1Reg);
-      killTemp1 = true;
-      }
-   // 64-bit needs sign extensions if we use the subf/addic/subfe sequence.
-   // therefore, we prefer the subf/cntlzw/rlwinm/xori sequence
-   if (cg->comp()->target().is64Bit())
-      {
-      generateTrg1Src1Instruction(cg, TR::InstOpCode::cntlzw, node, temp2Reg, temp1Reg);
-      generateShiftRightLogicalImmediate(cg, node, temp2Reg, temp2Reg, 5);
-      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::xori, node, trgReg, temp2Reg, 1);
-      }
-   else
-      {
-      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addic, node, temp2Reg, temp1Reg, -1);
-      generateTrg1Src2Instruction(cg, TR::InstOpCode::subfe, node, trgReg, temp2Reg, temp1Reg);
-      }
+   }
 
-   if (killTemp1)
-      cg->stopUsingRegister(temp1Reg);
-   cg->stopUsingRegister(temp2Reg);
+bool isSimpleSignedCompareToKnownSign(TR::Node *secondChild, const CompareInfo& compareInfo, TR::CodeGenerator *cg)
+   {
+   if (compareInfo.isUnsignedOrUnordered)
+      return false;
```

Comment:
why usigned int comparison not simple?  cmp and cmpl instructions have symmetrical flavors to me.

Reply:
It's not that unsigned comparisons aren't simple, it's that they aren't signed (and thus are not simple signed comparisons), which means that the optimization this routine is guarding does not work on them.